Add integration test to cover browser-side code

maxbrunsfeld opened the pull request:
This adds an integration test, `spec/integration/startup-spec.coffee` that tests Atom's handling of different command-line arguments. Currently, only a few simple cases are tested.

Questions:
- [x] Currently, `selenium-webdriver` has been added to the `dependencies`. We probably shouldn't require this module to be installed unless the user is running tests. Can we use `devDependencies`? _Moved the dependency to build/package.json_
- [x] Currently, the tests are skipped if `chromedriver` is not installed. Should we install `chromedriver` on CI? _We won't need to if we download it as part of the build; see below_
- [x] Should this test live in a separate suite so that it isn't run by default. The integration tests take about 14 seconds to run on my macbook pro. _Yes_
- [x] Is this even a good idea? I feel like something along these lines will help us in the future when developing the browser-side code. _Yes_
- [x] Make integration specs runnable separately from the main spec suite
  - [x] Make the integration specs _not_ run as part of the normal spec suite
  - [x] Make build script download `chromedriver` to the same location as `Atom.app`
  - [x] Make `run-specs` grunt task run the integration tests
  - [x] Publish new `grunt-download-atom-shell` version and upgrade it in atom
- [x] Fix the tests' flakiness

/cc @atom/core

maxbrunsfeld committed:
Add hidden --socket-path flag

This will allow integration tests to control which atom application
instance they use when creating atom windows

maxbrunsfeld committed:
Add integration test for starting atom w/ different arguments

nathansobo commented:
I think test coverage for this area of our system is probably a good idea. It will have a cost in terms of extra build complexity, but it will be great to have confidence in browser to render process interactions. I wonder if there's also a role for browser process unit tests so we don't end up testing everything browser-related with the integration tests. That's another day though.

kevinsawicki commented on src/browser/atom-application.coffee:
```diff
@@ -14,7 +14,7 @@ url = require 'url'
 _ = require 'underscore-plus'
 
-socketPath =
+DEFAULT_SOCKET_PATH =
```
I think we typically do `DefaultSocketPath` for things like this

kevinsawicki commented on package.json:
```diff
@@ -58,6 +58,7 @@
     "scoped-property-store": "^0.16.2",
     "scrollbar-style": "^2.0.0",
     "season": "^5.1.2",
+    "selenium-webdriver": "^2.44.0",
```
We don't want to ship this with Atom itself right?

If not, it should go in `build/package.json` then.

kevinsawicki commented on spec/integration/startup-spec.coffee:
```diff
@@ -0,0 +1,85 @@
+remote = require "remote"
+temp = require("temp").track()
+{spawn, spawnSync} = require "child_process"
+{Builder, By} = require("selenium-webdriver")
```
:fire: `()`

maxbrunsfeld committed:
Move selenium-webdriver dependency to build/package.json

maxbrunsfeld committed:
:art: Rename DEFAULT_SOCKET_PATH -> DefaultSocketPath

maxbrunsfeld committed:
Add integration test coverage for reusing windows w/ the same dir

maxbrunsfeld committed:
Use varargs in integration-test helper fn

maxbrunsfeld committed:
Use which(1), not bash's type function

maxbrunsfeld committed:
Download chromedriver along with atom-shell

maxbrunsfeld committed:
Enable integration tests with environment var

I think this makes more sense then running them via a separate command,
since that command would only make sense for atom-core, unlike the
current 'run-package-specs' command, which works for any atom package.

This way, they won't run by default, but you can opt in to running them
on the command line by setting an env var, or in the spec runner by
temporarily editing the code, like we do for focused tests anyway.

maxbrunsfeld committed:
Run integration tests on CI

maxbrunsfeld committed:
:arrow_up: grunt-download-atom-shell

maxbrunsfeld committed:
Wait for chromedriver's startup message on stdout

maxbrunsfeld committed:
Use webdriverio, not selenium-webdriver

maxbrunsfeld commented:
Passed on CI after fixing some flaky parts. Gonna re-run it repeatedly before merging.

maxbrunsfeld commented:
Alright 4 in a row. After 8, I'm gonna :ship:.

mark-hahn commented:
That ship icon is not legible.  I have to blow up my page each time to see
it.  Of course my eyes aren't great.

On Fri, Feb 6, 2015 at 9:51 PM, Max Brunsfeld <notifications@github.com>
wrote:

> Alright 4 in a row. After 8, I'm gonna [image: :ship:].
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/atom/atom/pull/5404#issuecomment-73351444>.
>

lee-dohm commented:
@mark-hahn You can always just hover over the emoji to get the text equivalent.

maxbrunsfeld committed:
Specify ATOM_HOME when starting Atom in integration test

Otherwise, tests may fail because the Welcome pane may appear,
depending on the content of `~/.atom/config.cson`

maxbrunsfeld committed:
Put socket in new temp dir on every test run

maxbrunsfeld commented:
Alright, so several package specs have been failing intermittently on CI lately, but the core specs have now passed 12 times in a row, so I feel that they're ready to go.

kevinsawicki commented on build/tasks/spec-task.coffee:
```diff
@@ -85,15 +85,27 @@ module.exports = (grunt) ->
+        opts:
+          env: _.extend({}, process.env,
+            ATOM_INTEGRATION_TESTS_ENABLED: true
+            PATH: [process.env.path, chromedriverPath].join(":")
```
The `path.delimiter` constant is probably better here than hard-coding `':'`

nathansobo commented:
Thanks for taking the initiative and expanding our test coverage.
